### ISSUE TITLE
Knowledge Pool Issue

### ISSUE THREAD

**Issue by CountAndromalius** [magefree/mage#4947]
Cards cast from other zones that aren't the hand should not trigger Knowledge Pool, as it states that only cards cast from the hand should be exiled afterwards.

Example: cards coming off suspend shouldn't trigger Knowledge Pool.

**Comment by Kurolox** [comment]
I was part of the game where we had this issue. The specific scenario was a Rift Bolt coming out of suspend, but wasn't able to cast it because of knowledge pool.

Here's the log file from the match. It happened pretty much at the end.
```
<html>
  <head>
    
  </head>
  <body>
    <font color="#CCCC33">10:23 PM: </font><font color="#20B2AA">Kurolox</font><font color="Orange"> 
    has joined the game</font>
<font color="#CCCC33">10:23 PM: </font><font color="#20B2AA">dead</font><font color="Orange"> 
    has joined the game</font>
<font color="Orange"><br><b>Match score:</b><br></font><font color="#20B2AA">dead</font><font color="Orange"> 
    - 0 wins<br>DeckHash: -7109478687083413595<br></font><font color="#20B2AA">Kurolox</font><font color="Orange"> 
    - 0 wins<br>DeckHash: -8355550088082251541<br>Range: ONE<br><br>Match is 
    not rated<br>You have to win 2 games to win the complete match<br><br>Game 
    has started<br><br></font>
<font color="#CCCC33">10:23 PM: </font><font color="#20B2AA">Kurolox</font><font color="White">'s 
    library is shuffled</font>
<font color="#CCCC33">10:23 PM: </font><font color="#20B2AA">dead</font><font color="White">'s 
    library is shuffled</font>
<font color="#CCCC33">10:23 PM: </font><font color="#20B2AA">Kurolox</font><font color="White"> 
    won the toss</font>
<font color="#CCCC33">10:23 PM: </font><font color="#20B2AA">Kurolox</font><font color="White"> 
    chooses that he or she takes the first turn</font>
<font color="#CCCC33">10:23 PM: </font><font color="#20B2AA">Kurolox</font><font color="White"> 
    draws seven cards</font>
<font color="#CCCC33">10:23 PM: </font><font color="#20B2AA">dead</font><font color="White"> 
    draws seven cards</font>
<font color="#CCCC33">10:23 PM: </font><font color="#20B2AA">Kurolox</font><font color="White"> 
    decides to take mulligan</font>
<font color="#CCCC33">10:23 PM: </font><font color="#20B2AA">dead</font><font color="White"> 
    decides to take mulligan</font>
<font color="#CCCC33">10:23 PM: </font><font color="#20B2AA">Kurolox</font><font color="White">'s 
    library is shuffled</font>
<font color="#CCCC33">10:23 PM: </font><font color="#20B2AA">Kurolox</font><font color="White"> 
    mulligans down to 6 cards</font>
<font color="#CCCC33">10:23 PM: </font><font color="#20B2AA">Kurolox</font><font color="White"> 
    draws six cards</font>
<font color="#CCCC33">10:23 PM: </font><font color="#20B2AA">dead</font><font color="White">'s 
    library is shuffled</font>
<font color="#CCCC33">10:23 PM: </font><font color="#20B2AA">dead</font><font color="White"> 
    mulligans down to 6 cards</font>
<font color="#CCCC33">10:23 PM: </font><font color="#20B2AA">dead</font><font color="White"> 
    draws six cards</font>
<font color="#CCCC33">10:23 PM: </font><font color="#20B2AA">Kurolox</font><font color="White"> 
    decides to take mulligan</font>
<font color="#CCCC33">10:23 PM: </font><font color="#20B2AA">dead</font><font color="White"> 
    decides to take mulligan</font>
<font color="#CCCC33">10:23 PM: </font><font color="#20B2AA">Kurolox</font><font color="White">'s 
    library is shuffled</font>
<font color="#CCCC33">10:23 PM: </font><font color="#20B2AA">Kurolox</font><font color="White"> 
    mulligans down to 5 cards</font>
<font color="#CCCC33">10:23 PM: </font><font color="#20B2AA">Kurolox</font><font color="White"> 
    draws five cards</font>
<font color="#CCCC33">10:23 PM: </font><font color="#20B2AA">dead</font><font color="White">'s 
    library is shuffled</font>
<font color="#CCCC33">10:23 PM: </font><font color="#20B2AA">dead</font><font color="White"> 
    mulligans down to 5 cards</font>
<font color="#CCCC33">10:23 PM: </font><font color="#20B2AA">dead</font><font color="White"> 
    draws five cards</font>
<font color="#CCCC33">10:23 PM: </font><font color="#20B2AA">Kurolox</font><font color="White"> 
    keeps hand</font>
<font color="#CCCC33">10:23 PM: </font><font color="#20B2AA">dead</font><font color="White"> 
    keeps hand</font>
<font color="#CCCC33">10:23 PM: </font><font color="#20B2AA">Kurolox</font><font color="White"> 
    scries 1</font>
<font color="#CCCC33">10:23 PM: </font><font color="#20B2AA">Kurolox</font><font color="White"> 
    puts a card from library to the top of their library</font>
<font color="#CCCC33">10:23 PM: </font><font color="#20B2AA">dead</font><font color="White"> 
    scries 1</font>
<font color="#CCCC33">10:23 PM: </font><font color="#20B2AA">dead</font><font color="White"> 
    puts a card from library to the top of their library</font>
<font color="#CCCC33">10:23 PM: </font><font color="Orange">Turn 1 </font><font color="#20B2AA">Kurolox</font><font color="Orange"> 
    (20 - 20)</font>
<font color="#CCCC33">10:23 PM: </font><font color="#20B2AA">Kurolox</font><font color="White"> 
    puts </font><font color="#B0C4DE"><a href="#Island">Island</a></font><font color="White"> 
    [ec7] from hand onto the Battlefield</font>
<font color="#CCCC33">10:23 PM: </font><font color="#20B2AA">Kurolox</font><font color="White"> 
    plays </font><font color="#B0C4DE"><a href="#Island">Island</a></font><font color="White"> 
    [ec7]</font>
<font color="#CCCC33">10:23 PM: </font><font color="#20B2AA">Kurolox</font><font color="White"> 
    attacks with 0 creatures</font>
<font color="#CCCC33">10:23 PM: </font><font color="Orange">Turn 2 </font><font color="#20B2AA">dead</font><font color="Orange"> 
    (20 - 20)</font>
<font color="#CCCC33">10:23 PM: </font><font color="#20B2AA">dead</font><font color="White"> 
    draws a card</font>
<font color="#CCCC33">10:24 PM: </font><font color="#20B2AA">dead</font><font color="White"> 
    puts </font><font color="#B0C4DE"><a href="#Flooded Strand">Flooded Strand</a></font><font color="White"> 
    [c5f] from hand onto the Battlefield</font>
<font color="#CCCC33">10:24 PM: </font><font color="#20B2AA">dead</font><font color="White"> 
    plays </font><font color="#B0C4DE"><a href="#Flooded Strand">Flooded Strand</a></font><font color="White"> 
    [c5f]</font>
<font color="#CCCC33">10:24 PM: </font><font color="#20B2AA">dead</font><font color="White"> 
    attacks with 0 creatures</font>
<font color="#CCCC33">10:24 PM: </font><font color="Orange">Turn 3 </font><font color="#20B2AA">Kurolox</font><font color="Orange"> 
    (20 - 20)</font>
<font color="#CCCC33">10:24 PM: </font><font color="#20B2AA">Kurolox</font><font color="White"> 
    draws a card</font>
<font color="#CCCC33">10:24 PM: </font><font color="#20B2AA">Kurolox</font><font color="White"> 
    casts </font><font color="#87CEFA"><a href="#Opt">Opt</a></font><font color="White"> 
    [e33]</font>
<font color="#CCCC33">10:24 PM: </font><font color="#20B2AA">Kurolox</font><font color="White"> 
    scries 1</font>
<font color="#CCCC33">10:24 PM: </font><font color="#20B2AA">Kurolox</font><font color="White"> 
    puts a card from library to the bottom of their library</font>
<font color="#CCCC33">10:24 PM: </font><font color="#20B2AA">Kurolox</font><font color="White"> 
    draws a card</font>
<font color="#CCCC33">10:24 PM: </font><font color="#20B2AA">Kurolox</font><font color="White"> 
    puts </font><font color="#87CEFA"><a href="#Opt">Opt</a></font><font color="White"> 
    [e33] from stack into their graveyard</font>
<font color="#CCCC33">10:24 PM: </font><font color="#20B2AA">Kurolox</font><font color="White"> 
    puts </font><font color="#B0C4DE"><a href="#Mountain">Mountain</a></font><font color="White"> 
    [85f] from hand onto the Battlefield</font>
<font color="#CCCC33">10:24 PM: </font><font color="#20B2AA">Kurolox</font><font color="White"> 
    plays </font><font color="#B0C4DE"><a href="#Mountain">Mountain</a></font><font color="White"> 
    [85f]</font>
<font color="#CCCC33">10:24 PM: </font><font color="#20B2AA">Kurolox</font><font color="White"> 
    casts </font><font color="#FF6347"><a href="#Soul-Scar Mage">Soul-Scar Mage</a></font><font color="White"> 
    [202]</font>
<font color="#CCCC33">10:24 PM: </font><font color="#20B2AA">Kurolox</font><font color="White"> 
    puts </font><font color="#FF6347"><a href="#Soul-Scar Mage">Soul-Scar Mage</a></font><font color="White"> 
    [202] from stack onto the Battlefield</font>
<font color="#CCCC33">10:24 PM: </font><font color="#20B2AA">Kurolox</font><font color="White"> 
    attacks with 0 creatures</font>
<font color="#CCCC33">10:24 PM: </font><font color="#20B2AA">dead</font><font color="White"> 
    loses 1 life</font>
<font color="#CCCC33">10:24 PM: </font><font color="#20B2AA">dead</font><font color="White"> 
    sacrificed </font><font color="#B0C4DE"><a href="#Flooded Strand">Flooded 
    Strand</a></font><font color="White"> [c5f]</font>
<font color="#CCCC33">10:24 PM: </font><font color="#20B2AA">dead</font><font color="White"> 
    activates: search your library for a Island or Plains and put it onto the 
    battlefield. Then shuffle your library. from </font><font color="#B0C4DE"><a href="#Flooded Strand">Flooded 
    Strand</a></font><font color="White"> [c5f]</font>
<font color="#CCCC33">10:24 PM: </font><font color="#20B2AA">dead</font><font color="White"> 
    searches their library</font>
<font color="#CCCC33">10:24 PM: </font><font color="#20B2AA">dead</font><font color="White"> 
    puts </font><font color="#B0C4DE"><a href="#Hallowed Fountain">Hallowed 
    Fountain</a></font><font color="White"> [a4c] from library onto the 
    Battlefield</font>
<font color="#CCCC33">10:24 PM: </font><font color="#20B2AA">dead</font><font color="White">'s 
    library is shuffled</font>
<font color="#CCCC33">10:24 PM: </font><font color="Orange">Turn 4 </font><font color="#20B2AA">dead</font><font color="Orange"> 
    (20 - 19)</font>
<font color="#CCCC33">10:24 PM: </font><font color="#20B2AA">dead</font><font color="White"> 
    draws a card</font>
<font color="#CCCC33">10:24 PM: </font><font color="#20B2AA">dead</font><font color="White"> 
    puts </font><font color="#B0C4DE"><a href="#Plains">Plains</a></font><font color="White"> 
    [5a9] from hand onto the Battlefield</font>
<font color="#CCCC33">10:24 PM: </font><font color="#20B2AA">dead</font><font color="White"> 
    plays </font><font color="#B0C4DE"><a href="#Plains">Plains</a></font><font color="White"> 
    [5a9]</font>
<font color="#CCCC33">10:24 PM: </font><font color="#20B2AA">dead</font><font color="White"> 
    casts </font><font color="#B0C4DE"><a href="#Azorius Signet">Azorius Signet</a></font><font color="White"> 
    [b8d]</font>
<font color="#CCCC33">10:24 PM: </font><font color="#20B2AA">dead</font><font color="White"> 
    puts </font><font color="#B0C4DE"><a href="#Azorius Signet">Azorius Signet</a></font><font color="White"> 
    [b8d] from stack onto the Battlefield</font>
<font color="#CCCC33">10:24 PM: </font><font color="#20B2AA">dead</font><font color="White"> 
    attacks with 0 creatures</font>
<font color="#CCCC33">10:24 PM: </font><font color="Orange">Turn 5 </font><font color="#20B2AA">Kurolox</font><font color="Orange"> 
    (20 - 19)</font>
<font color="#CCCC33">10:24 PM: </font><font color="#20B2AA">Kurolox</font><font color="White"> 
    draws a card</font>
<font color="#CCCC33">10:24 PM: </font><font color="#20B2AA">Kurolox</font><font color="White"> 
    puts </font><font color="#B0C4DE"><a href="#Mountain">Mountain</a></font><font color="White"> 
    [be0] from hand onto the Battlefield</font>
<font color="#CCCC33">10:24 PM: </font><font color="#20B2AA">Kurolox</font><font color="White"> 
    plays </font><font color="#B0C4DE"><a href="#Mountain">Mountain</a></font><font color="White"> 
    [be0]</font>
<font color="#CCCC33">10:24 PM: </font><font color="#20B2AA">Kurolox</font><font color="White"> 
    casts </font><font color="#DAA520"><a href="#Stormchaser Mage">Stormchaser 
    Mage</a></font><font color="White"> [89c]</font>
<font color="#CCCC33">10:25 PM: </font><font color="#20B2AA">Kurolox</font><font color="White"> 
    puts </font><font color="#DAA520"><a href="#Stormchaser Mage">Stormchaser 
    Mage</a></font><font color="White"> [89c] from stack onto the Battlefield</font>
<font color="#CCCC33">10:25 PM: </font><font color="#20B2AA">Kurolox</font><font color="White"> 
    attacks with 2 creatures</font>
<font color="#CCCC33">10:25 PM: </font><font color="#20B2AA">Kurolox</font><font color="White"> 
    casts </font><font color="#FF6347"><a href="#Lightning Bolt">Lightning Bolt</a></font><font color="White"> 
    [196] targeting </font><font color="#20B2AA">dead</font><font color="White"> 
    </font>
<font color="#CCCC33">10:25 PM: </font><font color="White">Ability triggers: </font><font color="#DAA520"><a href="#Stormchaser Mage">Stormchaser 
    Mage</a></font><font color="White"> [89c] - Prowess <i>(Whenever you cast 
    a noncreature spell, this creature gets +1/+1 until end of turn.)</i></font>
<font color="#CCCC33">10:25 PM: </font><font color="White">Ability triggers: </font><font color="#FF6347"><a href="#Soul-Scar Mage">Soul-Scar 
    Mage</a></font><font color="White"> [202] - Prowess <i>(Whenever you cast 
    a noncreature spell, this creature gets +1/+1 until end of turn.)</i></font>
<font color="#CCCC33">10:25 PM: </font><font color="#20B2AA">dead</font><font color="White"> 
    loses 3 life</font>
<font color="#CCCC33">10:25 PM: </font><font color="#20B2AA">Kurolox</font><font color="White"> 
    puts </font><font color="#FF6347"><a href="#Lightning Bolt">Lightning Bolt</a></font><font color="White"> 
    [196] from stack into their graveyard</font>
<font color="#CCCC33">10:25 PM: </font><font color="White">Attacker: </font><font color="#FF6347"><a href="#Soul-Scar Mage">Soul-Scar 
    Mage</a></font><font color="White"> [202] (2/3) unblocked</font>
<font color="#CCCC33">10:25 PM: </font><font color="White">Attacker: </font><font color="#DAA520"><a href="#Stormchaser Mage">Stormchaser 
    Mage</a></font><font color="White"> [89c] (2/4) unblocked</font>
<font color="#CCCC33">10:25 PM: </font><font color="#20B2AA">dead</font><font color="White"> 
    loses 2 life</font>
<font color="#CCCC33">10:25 PM: </font><font color="#20B2AA">dead</font><font color="White"> 
    loses 2 life</font>
<font color="#CCCC33">10:25 PM: </font><font color="Orange">Turn 6 </font><font color="#20B2AA">dead</font><font color="Orange"> 
    (20 - 12)</font>
<font color="#CCCC33">10:25 PM: </font><font color="#20B2AA">dead</font><font color="White"> 
    draws a card</font>
<font color="#CCCC33">10:25 PM: </font><font color="#20B2AA">dead</font><font color="White"> 
    attacks with 0 creatures</font>
<font color="#CCCC33">10:25 PM: </font><font color="Orange">Turn 7 </font><font color="#20B2AA">Kurolox</font><font color="Orange"> 
    (20 - 12)</font>
<font color="#CCCC33">10:25 PM: </font><font color="#20B2AA">Kurolox</font><font color="White"> 
    draws a card</font>
<font color="#CCCC33">10:25 PM: </font><font color="#20B2AA">Kurolox</font><font color="White"> 
    casts </font><font color="#FF6347"><a href="#Young Pyromancer">Young 
    Pyromancer</a></font><font color="White"> [9c6]</font>
<font color="#CCCC33">10:25 PM: </font><font color="#20B2AA">Kurolox</font><font color="White"> 
    puts </font><font color="#FF6347"><a href="#Young Pyromancer">Young 
    Pyromancer</a></font><font color="White"> [9c6] from stack onto the 
    Battlefield</font>
<font color="#CCCC33">10:26 PM: </font><font color="#20B2AA">Kurolox</font><font color="White"> 
    attacks with 2 creatures</font>
<font color="#CCCC33">10:26 PM: </font><font color="White">Attacker: </font><font color="#FF6347"><a href="#Soul-Scar Mage">Soul-Scar 
    Mage</a></font><font color="White"> [202] (1/2) unblocked</font>
<font color="#CCCC33">10:26 PM: </font><font color="White">Attacker: </font><font color="#DAA520"><a href="#Stormchaser Mage">Stormchaser 
    Mage</a></font><font color="White"> [89c] (1/3) unblocked</font>
<font color="#CCCC33">10:26 PM: </font><font color="#20B2AA">dead</font><font color="White"> 
    loses 1 life</font>
<font color="#CCCC33">10:26 PM: </font><font color="#20B2AA">dead</font><font color="White"> 
    loses 1 life</font>
<font color="#CCCC33">10:26 PM: </font><font color="#20B2AA">dead</font><font color="White"> 
    casts </font><font color="#87CEFA"><a href="#Thirst for Knowledge">Thirst 
    for Knowledge</a></font><font color="White"> [c19]</font>
<font color="#CCCC33">10:26 PM: </font><font color="#20B2AA">dead</font><font color="White"> 
    draws three cards</font>
<font color="#CCCC33">10:26 PM: </font><font color="#20B2AA">dead</font><font color="White"> 
    discards </font><font color="#87CEFA"><a href="#Remand">Remand</a></font><font color="White"> 
    [cc6]</font>
<font color="#CCCC33">10:26 PM: </font><font color="#20B2AA">dead</font><font color="White"> 
    discards </font><font color="#87CEFA"><a href="#Remand">Remand</a></font><font color="White"> 
    [a0c]</font>
<font color="#CCCC33">10:26 PM: </font><font color="#20B2AA">dead</font><font color="White"> 
    puts </font><font color="#87CEFA"><a href="#Thirst for Knowledge">Thirst 
    for Knowledge</a></font><font color="White"> [c19] from stack into their 
    graveyard</font>
<font color="#CCCC33">10:26 PM: </font><font color="Orange">Turn 8 </font><font color="#20B2AA">dead</font><font color="Orange"> 
    (20 - 10)</font>
<font color="#CCCC33">10:26 PM: </font><font color="#20B2AA">dead</font><font color="White"> 
    draws a card</font>
<font color="#CCCC33">10:26 PM: </font><font color="#20B2AA">dead</font><font color="White"> 
    puts </font><font color="#B0C4DE"><a href="#Mystic Gate">Mystic Gate</a></font><font color="White"> 
    [04c] from hand onto the Battlefield</font>
<font color="#CCCC33">10:26 PM: </font><font color="#20B2AA">dead</font><font color="White"> 
    plays </font><font color="#B0C4DE"><a href="#Mystic Gate">Mystic Gate</a></font><font color="White"> 
    [04c]</font>
<font color="#CCCC33">10:27 PM: </font><font color="#20B2AA">dead</font><font color="White"> 
    casts </font><font color="#87CEFA"><a href="#Serum Visions">Serum Visions</a></font><font color="White"> 
    [cdf]</font>
<font color="#CCCC33">10:27 PM: </font><font color="#20B2AA">dead</font><font color="White"> 
    draws a card</font>
<font color="#CCCC33">10:27 PM: </font><font color="#20B2AA">dead</font><font color="White"> 
    scries 2</font>
<font color="#CCCC33">10:27 PM: </font><font color="#20B2AA">dead</font><font color="White"> 
    puts a card from library to the bottom of their library</font>
<font color="#CCCC33">10:27 PM: </font><font color="#20B2AA">dead</font><font color="White"> 
    puts a card from library to the bottom of their library</font>
<font color="#CCCC33">10:27 PM: </font><font color="#20B2AA">dead</font><font color="White"> 
    puts </font><font color="#87CEFA"><a href="#Serum Visions">Serum Visions</a></font><font color="White"> 
    [cdf] from stack into their graveyard</font>
<font color="#CCCC33">10:27 PM: </font><font color="#20B2AA">dead</font><font color="White"> 
    casts </font><font color="#87CEFA"><a href="#Serum Visions">Serum Visions</a></font><font color="White"> 
    [643]</font>
<font color="#CCCC33">10:27 PM: </font><font color="#20B2AA">dead</font><font color="White"> 
    draws a card</font>
<font color="#CCCC33">10:27 PM: </font><font color="#20B2AA">dead</font><font color="White"> 
    scries 2</font>
<font color="#CCCC33">10:27 PM: </font><font color="#20B2AA">dead</font><font color="White"> 
    puts a card from library to the top of their library</font>
<font color="#CCCC33">10:27 PM: </font><font color="#20B2AA">dead</font><font color="White"> 
    puts a card from library to the top of their library</font>
<font color="#CCCC33">10:27 PM: </font><font color="#20B2AA">dead</font><font color="White"> 
    puts </font><font color="#87CEFA"><a href="#Serum Visions">Serum Visions</a></font><font color="White"> 
    [643] from stack into their graveyard</font>
<font color="#CCCC33">10:27 PM: </font><font color="#20B2AA">dead</font><font color="White"> 
    attacks with 0 creatures</font>
<font color="#CCCC33">10:27 PM: </font><font color="Orange">Turn 9 </font><font color="#20B2AA">Kurolox</font><font color="Orange"> 
    (20 - 10)</font>
<font color="#CCCC33">10:27 PM: </font><font color="#20B2AA">Kurolox</font><font color="White"> 
    draws a card</font>
<font color="#CCCC33">10:27 PM: </font><font color="#20B2AA">Kurolox</font><font color="White"> 
    attacks with 3 creatures</font>
<font color="#CCCC33">10:27 PM: </font><font color="White">Attacker: </font><font color="#FF6347"><a href="#Soul-Scar Mage">Soul-Scar 
    Mage</a></font><font color="White"> [202] (1/2) unblocked</font>
<font color="#CCCC33">10:27 PM: </font><font color="White">Attacker: </font><font color="#DAA520"><a href="#Stormchaser Mage">Stormchaser 
    Mage</a></font><font color="White"> [89c] (1/3) unblocked</font>
<font color="#CCCC33">10:27 PM: </font><font color="White">Attacker: </font><font color="#FF6347"><a href="#Young Pyromancer">Young 
    Pyromancer</a></font><font color="White"> [9c6] (2/1) unblocked</font>
<font color="#CCCC33">10:28 PM: </font><font color="#20B2AA">Kurolox</font><font color="White"> 
    casts </font><font color="#FF6347"><a href="#Gut Shot">Gut Shot</a></font><font color="White"> 
    [488] targeting </font><font color="#20B2AA">dead</font><font color="White"> 
    </font>
<font color="#CCCC33">10:28 PM: </font><font color="White">Ability triggers: </font><font color="#FF6347"><a href="#Young Pyromancer">Young 
    Pyromancer</a></font><font color="White"> [9c6] - Whenever you cast 
    instant or sorcery spell, create a 1/1 red Elemental creature.</font>
<font color="#CCCC33">10:28 PM: </font><font color="White">Ability triggers: </font><font color="#DAA520"><a href="#Stormchaser Mage">Stormchaser 
    Mage</a></font><font color="White"> [89c] - Prowess <i>(Whenever you cast 
    a noncreature spell, this creature gets +1/+1 until end of turn.)</i></font>
<font color="#CCCC33">10:28 PM: </font><font color="White">Ability triggers: </font><font color="#FF6347"><a href="#Soul-Scar Mage">Soul-Scar 
    Mage</a></font><font color="White"> [202] - Prowess <i>(Whenever you cast 
    a noncreature spell, this creature gets +1/+1 until end of turn.)</i></font>
<font color="#CCCC33">10:28 PM: </font><font color="#20B2AA">Kurolox</font><font color="White"> 
    creates a </font><font color="#FF6347"><a href="#Elemental">Elemental</a></font><font color="White"> 
    [365] token</font>
<font color="#CCCC33">10:28 PM: </font><font color="#20B2AA">dead</font><font color="White"> 
    loses 1 life</font>
<font color="#CCCC33">10:28 PM: </font><font color="#20B2AA">Kurolox</font><font color="White"> 
    puts </font><font color="#FF6347"><a href="#Gut Shot">Gut Shot</a></font><font color="White"> 
    [488] from stack into their graveyard</font>
<font color="#CCCC33">10:28 PM: </font><font color="#20B2AA">dead</font><font color="White"> 
    casts </font><font color="#F0E68C"><a href="#Path to Exile">Path to Exile</a></font><font color="White"> 
    [ae5] targeting </font><font color="#DAA520"><a href="#Stormchaser Mage">Stormchaser 
    Mage</a></font><font color="White"> [89c] </font>
<font color="#CCCC33">10:28 PM: </font><font color="#20B2AA">dead</font><font color="White"> 
    moves </font><font color="#DAA520"><a href="#Stormchaser Mage">Stormchaser 
    Mage</a></font><font color="White"> [89c] from battlefield to the exile 
    zone</font>
<font color="#CCCC33">10:28 PM: </font><font color="#20B2AA">Kurolox</font><font color="White"> 
    searches their library</font>
<font color="#CCCC33">10:28 PM: </font><font color="#20B2AA">Kurolox</font><font color="White"> 
    puts </font><font color="#B0C4DE"><a href="#Mountain">Mountain</a></font><font color="White"> 
    [744] from library onto the Battlefield</font>
<font color="#CCCC33">10:28 PM: </font><font color="#20B2AA">Kurolox</font><font color="White">'s 
    library is shuffled</font>
<font color="#CCCC33">10:28 PM: </font><font color="#20B2AA">dead</font><font color="White"> 
    puts </font><font color="#F0E68C"><a href="#Path to Exile">Path to Exile</a></font><font color="White"> 
    [ae5] from stack into their graveyard</font>
<font color="#CCCC33">10:28 PM: </font><font color="#20B2AA">dead</font><font color="White"> 
    loses 2 life</font>
<font color="#CCCC33">10:28 PM: </font><font color="#20B2AA">dead</font><font color="White"> 
    loses 2 life</font>
<font color="#CCCC33">10:28 PM: </font><font color="Orange">Turn 10 </font><font color="#20B2AA">dead</font><font color="Orange"> 
    (20 - 5)</font>
<font color="#CCCC33">10:28 PM: </font><font color="#20B2AA">dead</font><font color="White"> 
    draws a card</font>
<font color="#CCCC33">10:29 PM: </font><font color="#20B2AA">dead</font><font color="White"> 
    casts </font><font color="#F0E68C"><a href="#Wrath of God">Wrath of God</a></font><font color="White"> 
    [385]</font>
<font color="#CCCC33">10:29 PM: </font><font color="#FF6347"><a href="#Soul-Scar Mage">Soul-Scar 
    Mage</a></font><font color="White"> [202] died</font>
<font color="#CCCC33">10:29 PM: </font><font color="#FF6347"><a href="#Young Pyromancer">Young 
    Pyromancer</a></font><font color="White"> [9c6] died</font>
<font color="#CCCC33">10:29 PM: </font><font color="#FF6347"><a href="#Elemental">Elemental</a></font><font color="White"> 
    [365] died</font>
<font color="#CCCC33">10:29 PM: </font><font color="#20B2AA">dead</font><font color="White"> 
    puts </font><font color="#F0E68C"><a href="#Wrath of God">Wrath of God</a></font><font color="White"> 
    [385] from stack into their graveyard</font>
<font color="#CCCC33">10:29 PM: </font><font color="#20B2AA">dead</font><font color="White"> 
    attacks with 0 creatures</font>
<font color="#CCCC33">10:29 PM: </font><font color="Orange">Turn 11 </font><font color="#20B2AA">Kurolox</font><font color="Orange"> 
    (20 - 5)</font>
<font color="#CCCC33">10:29 PM: </font><font color="#20B2AA">Kurolox</font><font color="White"> 
    draws a card</font>
<font color="#CCCC33">10:29 PM: </font><font color="#20B2AA">Kurolox</font><font color="White"> 
    attacks with 0 creatures</font>
<font color="#CCCC33">10:29 PM: </font><font color="Orange">Turn 12 </font><font color="#20B2AA">dead</font><font color="Orange"> 
    (20 - 5)</font>
<font color="#CCCC33">10:29 PM: </font><font color="#20B2AA">dead</font><font color="White"> 
    draws a card</font>
<font color="#CCCC33">10:29 PM: </font><font color="#20B2AA">dead</font><font color="White"> 
    puts </font><font color="#B0C4DE"><a href="#Island">Island</a></font><font color="White"> 
    [684] from hand onto the Battlefield</font>
<font color="#CCCC33">10:29 PM: </font><font color="#20B2AA">dead</font><font color="White"> 
    plays </font><font color="#B0C4DE"><a href="#Island">Island</a></font><font color="White"> 
    [684]</font>
<font color="#CCCC33">10:29 PM: </font><font color="#20B2AA">dead</font><font color="White"> 
    attacks with 0 creatures</font>
<font color="#CCCC33">10:29 PM: </font><font color="Orange">Turn 13 </font><font color="#20B2AA">Kurolox</font><font color="Orange"> 
    (20 - 5)</font>
<font color="#CCCC33">10:29 PM: </font><font color="#20B2AA">Kurolox</font><font color="White"> 
    draws a card</font>
<font color="#CCCC33">10:29 PM: </font><font color="#20B2AA">Kurolox</font><font color="White"> 
    casts </font><font color="#87CEFA"><a href="#Opt">Opt</a></font><font color="White"> 
    [af5]</font>
<font color="#CCCC33">10:29 PM: </font><font color="#20B2AA">Kurolox</font><font color="White"> 
    scries 1</font>
<font color="#CCCC33">10:30 PM: </font><font color="#20B2AA">Kurolox</font><font color="White"> 
    puts a card from library to the bottom of their library</font>
<font color="#CCCC33">10:30 PM: </font><font color="#20B2AA">Kurolox</font><font color="White"> 
    draws a card</font>
<font color="#CCCC33">10:30 PM: </font><font color="#20B2AA">Kurolox</font><font color="White"> 
    puts </font><font color="#87CEFA"><a href="#Opt">Opt</a></font><font color="White"> 
    [af5] from stack into their graveyard</font>
<font color="#CCCC33">10:30 PM: </font><font color="#20B2AA">Kurolox</font><font color="White"> 
    attacks with 0 creatures</font>
<font color="#CCCC33">10:30 PM: </font><font color="#20B2AA">dead</font><font color="White"> 
    casts </font><font color="#87CEFA"><a href="#Teferi, Mage of Zhalfir">Teferi, 
    Mage of Zhalfir</a></font><font color="White"> [9d1]</font>
<font color="#CCCC33">10:30 PM: </font><font color="#20B2AA">dead</font><font color="White"> 
    puts </font><font color="#87CEFA"><a href="#Teferi, Mage of Zhalfir">Teferi, 
    Mage of Zhalfir</a></font><font color="White"> [9d1] from stack onto the 
    Battlefield</font>
<font color="#CCCC33">10:30 PM: </font><font color="Orange">Turn 14 </font><font color="#20B2AA">dead</font><font color="Orange"> 
    (20 - 5)</font>
<font color="#CCCC33">10:30 PM: </font><font color="#20B2AA">dead</font><font color="White"> 
    draws a card</font>
<font color="#CCCC33">10:30 PM: </font><font color="#20B2AA">dead</font><font color="White"> 
    casts </font><font color="#87CEFA"><a href="#Serum Visions">Serum Visions</a></font><font color="White"> 
    [54a]</font>
<font color="#CCCC33">10:30 PM: </font><font color="#20B2AA">dead</font><font color="White"> 
    draws a card</font>
<font color="#CCCC33">10:30 PM: </font><font color="#20B2AA">dead</font><font color="White"> 
    scries 2</font>
<font color="#CCCC33">10:30 PM: </font><font color="#20B2AA">dead</font><font color="White"> 
    puts a card from library to the bottom of their library</font>
<font color="#CCCC33">10:30 PM: </font><font color="#20B2AA">dead</font><font color="White"> 
    puts a card from library to the bottom of their library</font>
<font color="#CCCC33">10:30 PM: </font><font color="#20B2AA">dead</font><font color="White"> 
    puts </font><font color="#87CEFA"><a href="#Serum Visions">Serum Visions</a></font><font color="White"> 
    [54a] from stack into their graveyard</font>
<font color="#CCCC33">10:30 PM: </font><font color="#20B2AA">dead</font><font color="White"> 
    casts </font><font color="#B0C4DE"><a href="#Azorius Signet">Azorius Signet</a></font><font color="White"> 
    [f54]</font>
<font color="#CCCC33">10:30 PM: </font><font color="#20B2AA">dead</font><font color="White"> 
    puts </font><font color="#B0C4DE"><a href="#Azorius Signet">Azorius Signet</a></font><font color="White"> 
    [f54] from stack onto the Battlefield</font>
<font color="#CCCC33">10:31 PM: </font><font color="#20B2AA">dead</font><font color="White"> 
    attacks with 0 creatures</font>
<font color="#CCCC33">10:31 PM: </font><font color="Orange">Turn 15 </font><font color="#20B2AA">Kurolox</font><font color="Orange"> 
    (20 - 5)</font>
<font color="#CCCC33">10:31 PM: </font><font color="#20B2AA">Kurolox</font><font color="White"> 
    draws a card</font>
<font color="#CCCC33">10:31 PM: </font><font color="#20B2AA">Kurolox</font><font color="White"> 
    puts </font><font color="#B0C4DE"><a href="#Island">Island</a></font><font color="White"> 
    [891] from hand onto the Battlefield</font>
<font color="#CCCC33">10:31 PM: </font><font color="#20B2AA">Kurolox</font><font color="White"> 
    plays </font><font color="#B0C4DE"><a href="#Island">Island</a></font><font color="White"> 
    [891]</font>
<font color="#CCCC33">10:31 PM: </font><font color="#20B2AA">Kurolox</font><font color="White"> 
    casts </font><font color="#87CEFA"><a href="#Opt">Opt</a></font><font color="White"> 
    [e78]</font>
<font color="#CCCC33">10:31 PM: </font><font color="#20B2AA">Kurolox</font><font color="White"> 
    scries 1</font>
<font color="#CCCC33">10:31 PM: </font><font color="#20B2AA">Kurolox</font><font color="White"> 
    puts a card from library to the bottom of their library</font>
<font color="#CCCC33">10:31 PM: </font><font color="#20B2AA">Kurolox</font><font color="White"> 
    draws a card</font>
<font color="#CCCC33">10:31 PM: </font><font color="#20B2AA">Kurolox</font><font color="White"> 
    puts </font><font color="#87CEFA"><a href="#Opt">Opt</a></font><font color="White"> 
    [e78] from stack into their graveyard</font>
<font color="#CCCC33">10:31 PM: </font><font color="#20B2AA">Kurolox</font><font color="White"> 
    casts </font><font color="#FF6347"><a href="#Young Pyromancer">Young 
    Pyromancer</a></font><font color="White"> [b9d]</font>
<font color="#CCCC33">10:31 PM: </font><font color="#20B2AA">Kurolox</font><font color="White"> 
    puts </font><font color="#FF6347"><a href="#Young Pyromancer">Young 
    Pyromancer</a></font><font color="White"> [b9d] from stack onto the 
    Battlefield</font>
<font color="#CCCC33">10:31 PM: </font><font color="#20B2AA">Kurolox</font><font color="White"> 
    activates: Suspend 1. from </font><font color="#FF6347"><a href="#Rift Bolt">Rift 
    Bolt</a></font><font color="White"> [cb9]</font>
<font color="#CCCC33">10:31 PM: </font><font color="#20B2AA">Kurolox</font><font color="White"> 
    moves </font><font color="#FF6347"><a href="#Rift Bolt">Rift Bolt</a></font><font color="White"> 
    [cb9] from hand to the exile zone</font>
<font color="#CCCC33">10:31 PM: </font><font color="#20B2AA">Kurolox</font><font color="White"> 
    suspends (1) </font><font color="#FF6347"><a href="#Rift Bolt">Rift Bolt</a></font><font color="White"> 
    [cb9]</font>
<font color="#CCCC33">10:31 PM: </font><font color="#20B2AA">Kurolox</font><font color="White"> 
    attacks with 0 creatures</font>
<font color="#CCCC33">10:31 PM: </font><font color="#20B2AA">dead</font><font color="White"> 
    casts </font><font color="#87CEFA"><a href="#Thirst for Knowledge">Thirst 
    for Knowledge</a></font><font color="White"> [0a8]</font>
<font color="#CCCC33">10:31 PM: </font><font color="#20B2AA">dead</font><font color="White"> 
    draws three cards</font>
<font color="#CCCC33">10:31 PM: </font><font color="#20B2AA">dead</font><font color="White"> 
    discards </font><font color="#DAA520"><a href="#Detention Sphere">Detention 
    Sphere</a></font><font color="White"> [b28]</font>
<font color="#CCCC33">10:31 PM: </font><font color="#20B2AA">dead</font><font color="White"> 
    discards </font><font color="#87CEFA"><a href="#Teferi, Mage of Zhalfir">Teferi, 
    Mage of Zhalfir</a></font><font color="White"> [a5f]</font>
<font color="#CCCC33">10:31 PM: </font><font color="#20B2AA">dead</font><font color="White"> 
    puts </font><font color="#87CEFA"><a href="#Thirst for Knowledge">Thirst 
    for Knowledge</a></font><font color="White"> [0a8] from stack into their 
    graveyard</font>
<font color="#CCCC33">10:31 PM: </font><font color="Orange">Turn 16 </font><font color="#20B2AA">dead</font><font color="Orange"> 
    (20 - 5)</font>
<font color="#CCCC33">10:31 PM: </font><font color="#20B2AA">dead</font><font color="White"> 
    draws a card</font>
<font color="#CCCC33">10:31 PM: </font><font color="#20B2AA">dead</font><font color="White"> 
    puts </font><font color="#B0C4DE"><a href="#Island">Island</a></font><font color="White"> 
    [6c6] from hand onto the Battlefield</font>
<font color="#CCCC33">10:31 PM: </font><font color="#20B2AA">dead</font><font color="White"> 
    plays </font><font color="#B0C4DE"><a href="#Island">Island</a></font><font color="White"> 
    [6c6]</font>
<font color="#CCCC33">10:32 PM: </font><font color="#20B2AA">dead</font><font color="White"> 
    casts </font><font color="#B0C4DE"><a href="#Knowledge Pool">Knowledge Pool</a></font><font color="White"> 
    [f0e]</font>
<font color="#CCCC33">10:32 PM: </font><font color="#20B2AA">dead</font><font color="White"> 
    puts </font><font color="#B0C4DE"><a href="#Knowledge Pool">Knowledge Pool</a></font><font color="White"> 
    [f0e] from stack onto the Battlefield</font>
<font color="#CCCC33">10:32 PM: </font><font color="White">Ability triggers: </font><font color="#B0C4DE"><a href="#Knowledge Pool">Knowledge 
    Pool</a></font><font color="White"> [f0e] - When </font><font color="#B0C4DE"><a href="#Knowledge Pool">Knowledge 
    Pool</a></font><font color="White"> [f0e] enters the battlefield, each 
    player exiles the top three cards of their library.</font>
<font color="#CCCC33">10:32 PM: </font><font color="#20B2AA">dead</font><font color="White"> 
    moves </font><font color="#B0C4DE"><a href="#Field of Ruin">Field of Ruin</a></font><font color="White"> 
    [434] from library to the exile zone</font>
<font color="#CCCC33">10:32 PM: </font><font color="#20B2AA">dead</font><font color="White"> 
    moves </font><font color="#F0E68C"><a href="#Path to Exile">Path to Exile</a></font><font color="White"> 
    [f7d] from library to the exile zone</font>
<font color="#CCCC33">10:32 PM: </font><font color="#20B2AA">dead</font><font color="White"> 
    moves </font><font color="#B0C4DE"><a href="#Glacial Fortress">Glacial 
    Fortress</a></font><font color="White"> [693] from library to the exile 
    zone</font>
<font color="#CCCC33">10:32 PM: </font><font color="#20B2AA">Kurolox</font><font color="White"> 
    moves </font><font color="#B0C4DE"><a href="#Island">Island</a></font><font color="White"> 
    [bbe] from library to the exile zone</font>
<font color="#CCCC33">10:32 PM: </font><font color="#20B2AA">Kurolox</font><font color="White"> 
    moves </font><font color="#87CEFA"><a href="#Serum Visions">Serum Visions</a></font><font color="White"> 
    [48c] from library to the exile zone</font>
<font color="#CCCC33">10:32 PM: </font><font color="#20B2AA">Kurolox</font><font color="White"> 
    moves </font><font color="#90EE90"><a href="#Mutagenic Growth">Mutagenic 
    Growth</a></font><font color="White"> [a02] from library to the exile zone</font>
<font color="#CCCC33">10:32 PM: </font><font color="#20B2AA">dead</font><font color="White"> 
    attacks with 0 creatures</font>
<font color="#CCCC33">10:32 PM: </font><font color="Orange">Turn 17 </font><font color="#20B2AA">Kurolox</font><font color="Orange"> 
    (20 - 5)</font>
<font color="#CCCC33">10:32 PM: </font><font color="White">Ability triggers: </font><font color="#FF6347"><a href="#Rift Bolt">Rift 
    Bolt</a></font><font color="White"> [cb9] - At the beginning of your 
    upkeep, if this card (</font><font color="#FF6347"><a href="#Rift Bolt">Rift 
    Bolt</a></font><font color="White"> [cb9]) is suspended, remove a time 
    counter from it.</font>
<font color="#CCCC33">10:32 PM: </font><font color="White">Removed 1 time 
    counter from </font><font color="#FF6347"><a href="#Rift Bolt">Rift Bolt</a></font><font color="White"> 
    [cb9] (0 left)</font>
<font color="#CCCC33">10:32 PM: </font><font color="White">Ability triggers: </font><font color="#FF6347"><a href="#Rift Bolt">Rift 
    Bolt</a></font><font color="White"> [cb9] - When the last time counter is 
    removed from this card (</font><font color="#FF6347"><a href="#Rift Bolt">Rift 
    Bolt</a></font><font color="White"> [cb9]), if it's removed from the game, 
    play it without paying its mana cost if able. If you can't, it remains 
    removed from the game.</font>
<font color="#CCCC33">10:32 PM: </font><font color="#20B2AA">Kurolox</font><font color="White"> 
    draws a card</font>
<font color="#CCCC33">10:32 PM: </font><font color="#20B2AA">Kurolox</font><font color="White"> 
    loses 2 life</font>
<font color="#CCCC33">10:32 PM: </font><font color="#20B2AA">Kurolox</font><font color="White"> 
    casts </font><font color="#90EE90"><a href="#Mutagenic Growth">Mutagenic 
    Growth</a></font><font color="White"> [ff7] targeting </font><font color="#FF6347"><a href="#Young Pyromancer">Young 
    Pyromancer</a></font><font color="White"> [b9d] </font>
<font color="#CCCC33">10:32 PM: </font><font color="White">Ability triggers: </font><font color="#FF6347"><a href="#Young Pyromancer">Young 
    Pyromancer</a></font><font color="White"> [b9d] - Whenever you cast 
    instant or sorcery spell, create a 1/1 red Elemental creature.</font>
<font color="#CCCC33">10:32 PM: </font><font color="White">Ability triggers: </font><font color="#B0C4DE"><a href="#Knowledge Pool">Knowledge 
    Pool</a></font><font color="White"> [f0e] - Whenever a player casts a 
    spell from their hand, that player exiles it. If the player does, he or 
    she may cast another nonland card exiled with </font><font color="#B0C4DE"><a href="#Knowledge Pool">Knowledge 
    Pool</a></font><font color="White"> [f0e] without paying that card's mana 
    cost.</font>
<font color="#CCCC33">10:32 PM: </font><font color="#20B2AA">dead</font><font color="White"> 
    moves </font><font color="#90EE90"><a href="#Mutagenic Growth">Mutagenic 
    Growth</a></font><font color="White"> [ff7] from stack to the exile zone</font>
<font color="#CCCC33">10:34 PM: </font><font color="#20B2AA">Kurolox</font><font color="White"> 
    has conceded.</font>
<font color="#CCCC33">10:34 PM: </font><font color="#20B2AA">Kurolox</font><font color="White"> 
    has lost the game.</font>
<font color="#CCCC33">10:34 PM: </font><font color="#20B2AA">dead</font><font color="White"> 
    has won the game</font>
  </body>
</html>
```

**Comment by LevelX2** [comment]
According to the log the Knowledge Pool triggered only for Mutagenic Growth and not for the Rift Bolt.
So probably there was another problem with the Rift Bolt.

**Comment by alejandrociatti** [comment]
It also lets you cast the same card you just played (It just has a nonland-card filter). 
Knowledge Pool states **another**. 

I'm willing to fix this card if someone points me in the right direction, I did not find the proper `Predicate` to filter  nonland + not the card just played / triggering K.P.

**Comment by JayDi85** [comment]
Plays from non hand (suspend) works fine -- pool not triggers. Plays another card was fixed via 6316f40914deab3ce6c2ba456dec6e59feb30c4a.

@alejandrociatti 
You can add ignored ids by runtime as `Predicates.not(new CardIdPredicate())`:
```
FilterNonlandCard realFilter = filter.copy();
realFilter.add(Predicates.not(new CardIdPredicate(spell.getSourceId())));
```
